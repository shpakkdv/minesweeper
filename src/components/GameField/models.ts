import { GameLevel } from 'constant';
import { Field } from 'models';
import { ActionCreator } from 'containers/GameField/models';

export interface GameFieldProps {
  field: Field;
  gameLevel: GameLevel;
  loading: boolean;
  startGame: ActionCreator.StartGame;
}
