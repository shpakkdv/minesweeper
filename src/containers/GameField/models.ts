import { GameLevel } from 'constant';
import { IAction, Field, Mines } from 'models';

export interface State {
  field: Field;
  mines: Mines;
}

export namespace Payload {
  export interface SetField {
    field: Field;
  }
  export interface SetMines {
    mines: Mines;
  }
  export interface SetMine {
    x: number;
    y: number;
    value: boolean;
  }
  export interface StartGame {
    level: GameLevel;
  }
}

export namespace Action {
  export type SetField = IAction<Payload.SetField>;
  export type SetMines = IAction<Payload.SetMines>;
  export type SetMine = IAction<Payload.SetMine>;
  export type StartGame = IAction<Payload.StartGame>;
}

export namespace ActionCreator {
  export type SetField = (field: Field) => Action.SetField;
  export type SetMines = (mines: Mines) => Action.SetMines;
  export type SetMine = (x: number, y: number, value: boolean) => Action.SetMine;
  export type StartGame = (level: GameLevel) => Action.StartGame;
}
