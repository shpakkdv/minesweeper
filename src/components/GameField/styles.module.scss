@import "styles/constants.scss";

.GameField {
  width: fit-content;
  height: fit-content;
  border: $fieldLinesWidth solid $fieldTableColor;
  background-color: $openCellsBorderColor;
  margin: auto;

  display: grid;
  grid-gap: $fieldLinesWidth;
}

.chooseLevel {
  display: flex;
  margin: 0 auto;
  flex-direction: column;
  font-size: 25px;

  & > button {
    width: 300px;
    height: 50px;
    margin-top: 30px;
    cursor: pointer;
  }
}

.option {
  width: fit-content;
  display: flex;
  justify-content: flex-start;
  align-items: center;

  & > * {
    cursor: pointer;
  }

  & > label {
    padding-left: 10px;
  }
}

.loading {
  margin: auto;
  font-size: 50px;
}
