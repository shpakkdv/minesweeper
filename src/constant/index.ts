// single constants
export const MINESWEEPER_URL = 'wss://hometask.eg1236.com/game1/';

// enums
export enum GameLevel {
  LEVEL_1 = '1',
  LEVEL_2 = '2',
  LEVEL_3 = '3',
  LEVEL_4 = '4',
}

export enum AppStatus {
  LOADING = 'loading',
  SOLVING = 'solving',
  FINISHED = 'finished',
}

// helpers
export const LevelLabelByLevel: Record<GameLevel, string> = {
  [GameLevel.LEVEL_1]: 'Easy',
  [GameLevel.LEVEL_2]: 'Medium',
  [GameLevel.LEVEL_3]: 'Hard',
  [GameLevel.LEVEL_4]: 'Impossible',
};

export const MinesNumberByLevel: Record<GameLevel, number> = {
  [GameLevel.LEVEL_1]: 15,
  [GameLevel.LEVEL_2]: 150,
  [GameLevel.LEVEL_3]: 1000,
  [GameLevel.LEVEL_4]: 3500,
};

export const TimeToSolveByLevel: Record<GameLevel, string> = {
  [GameLevel.LEVEL_1]: '30 seconds',
  [GameLevel.LEVEL_2]: '4 minutes',
  [GameLevel.LEVEL_3]: '15 minutes',
  [GameLevel.LEVEL_4]: 'one hour',
};

export const GAME_LEVELS: GameLevel[] = [GameLevel.LEVEL_1, GameLevel.LEVEL_2, GameLevel.LEVEL_3, GameLevel.LEVEL_4];
