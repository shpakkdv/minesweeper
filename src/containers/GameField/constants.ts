import { keyMirror } from 'utils/keyMirror';

export const ActionType = keyMirror(
  {
    SET_FIELD: null,
    SET_MINES: null,
    SET_MINE: null,
    START_GAME: null,
  },
  'GameField',
);
