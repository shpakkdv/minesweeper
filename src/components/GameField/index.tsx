import React, { useState } from 'react';

import Row from 'components/Row';
import { GameLevel, GAME_LEVELS, LevelLabelByLevel } from 'constant';
import { GameFieldProps } from './models';
import styles from './styles.module.scss';

const GameField: React.FunctionComponent<GameFieldProps> = ({ field, gameLevel, loading, startGame }) => {
  const [selectedLevel, setLevel] = useState(gameLevel ?? GAME_LEVELS[0]);

  const onStartGameClick: React.MouseEventHandler<HTMLButtonElement> = (): void => {
    startGame(selectedLevel);
  };

  if (loading) {
    return <div className={styles.loading}>Loading...</div>;
  }

  if (!gameLevel) {
    return (
      <div className={styles.chooseLevel}>
        <h2>Choose game level:</h2>
        <form>
          {GAME_LEVELS.map((level) => (
            <div key={level} className={styles.option}>
              {getRadioButton(level, LevelLabelByLevel[level], setLevel, selectedLevel === level)}
            </div>
          ))}
        </form>
        <button onClick={onStartGameClick}>START GAME</button>
      </div>
    );
  }

  return (
    <div className={styles.GameField} style={{ gridTemplateColumns: `repeat(${field[0].length}, 40px)`, gridAutoRows: '40px' }}>
      {field.map((row, index) => (
        <Row key={index} rowIndex={index} row={row} />
      ))}
    </div>
  );
};

export default GameField;

const getRadioButton = (id: GameLevel, label: string, onClick: (id: GameLevel) => void, checked: boolean): JSX.Element => (
  <>
    <input type="radio" id={id} name="level" onChange={(event) => event.target.checked && onClick(id)} checked={checked} />
    <label htmlFor={id}>{label}</label>
  </>
);
