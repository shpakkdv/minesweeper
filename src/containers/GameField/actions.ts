import { ActionType } from './constants';
import { ActionCreator } from './models';

export const setField: ActionCreator.SetField = (field) => ({
  type: ActionType.SET_FIELD,
  payload: {
    field,
  },
});

export const setMines: ActionCreator.SetMines = (mines) => ({
  type: ActionType.SET_MINES,
  payload: {
    mines,
  },
});

export const setMine: ActionCreator.SetMine = (x, y, value) => ({
  type: ActionType.SET_MINE,
  payload: {
    x,
    y,
    value,
  },
});

export const startGame: ActionCreator.StartGame = (level) => ({
  type: ActionType.START_GAME,
  payload: {
    level,
  },
});
