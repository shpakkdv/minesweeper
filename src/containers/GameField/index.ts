import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import GameField from 'components/GameField';
import { appStatus, gameLevel, cellSize } from 'containers/Controls/selectors';
import { startGame } from './actions';
import { field } from './selectors';

const mapStateToProps = createStructuredSelector({
  field,
  gameLevel,
  appStatus,
  cellSize,
});

const mapDispatchToProps = {
  startGame,
};

export default connect(mapStateToProps, mapDispatchToProps)(GameField);
