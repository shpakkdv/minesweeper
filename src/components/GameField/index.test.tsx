import React from 'react';
import { shallow } from 'enzyme';

import { AppStatus, GameLevel } from 'constant';
import GameField from './index';

describe('GameField', () => {
  it('Render game levels if the level is not chosen', () => {
    const wrapper = shallow(
      <GameField
        field={[[]]}
        gameLevel={null}
        appStatus={null}
        cellSize={40}
        startGame={jest.fn()}
      />,
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('input').length).toBeGreaterThan(1);
  });

  it('Render loading', () => {
    const wrapper = shallow(
      <GameField
        field={[[]]}
        gameLevel={GameLevel.LEVEL_1}
        appStatus={AppStatus.LOADING}
        cellSize={40}
        startGame={jest.fn()}
      />,
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.text()).toBe('Loading...');
  });

  it('Render the game', () => {
    const wrapper = shallow(
      <GameField
        field={[[-1, -1, -1], [-1, -1, -1], [-1, -1, -1]]}
        gameLevel={GameLevel.LEVEL_1}
        appStatus={null}
        cellSize={40}
        startGame={jest.fn()}
      />,
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('Row').length).toBe(3);
  });
});
