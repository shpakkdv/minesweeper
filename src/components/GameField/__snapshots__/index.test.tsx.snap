// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameField Render game levels if the level is not chosen 1`] = `
<div
  className="chooseLevel"
>
  <h2>
    Choose game level:
  </h2>
  <form>
    <div
      className="option"
      key="1"
    >
      <input
        checked={true}
        id="1"
        name="level"
        onChange={[Function]}
        type="radio"
      />
      <label
        htmlFor="1"
      >
        Easy
      </label>
    </div>
    <div
      className="option"
      key="2"
    >
      <input
        checked={false}
        id="2"
        name="level"
        onChange={[Function]}
        type="radio"
      />
      <label
        htmlFor="2"
      >
        Medium
      </label>
    </div>
    <div
      className="option"
      key="3"
    >
      <input
        checked={false}
        id="3"
        name="level"
        onChange={[Function]}
        type="radio"
      />
      <label
        htmlFor="3"
      >
        Hard
      </label>
    </div>
    <div
      className="option"
      key="4"
    >
      <input
        checked={false}
        id="4"
        name="level"
        onChange={[Function]}
        type="radio"
      />
      <label
        htmlFor="4"
      >
        Impossible
      </label>
    </div>
  </form>
  <button
    onClick={[Function]}
  >
    START GAME
  </button>
</div>
`;

exports[`GameField Render loading 1`] = `
<div
  className="loading"
>
  Loading...
</div>
`;

exports[`GameField Render the game 1`] = `
<div
  className="GameField"
  style={
    Object {
      "gridAutoRows": "40px",
      "gridTemplateColumns": "repeat(3, 40px)",
    }
  }
>
  <Row
    key="0"
    row={
      Array [
        -1,
        -1,
        -1,
      ]
    }
    rowIndex={0}
  />
  <Row
    key="1"
    row={
      Array [
        -1,
        -1,
        -1,
      ]
    }
    rowIndex={1}
  />
  <Row
    key="2"
    row={
      Array [
        -1,
        -1,
        -1,
      ]
    }
    rowIndex={2}
  />
</div>
`;
